rfc/rfc-83/rfc-83.md

```diff
@@ -0,0 +1,260 @@
+
+public interface TableServiceCommitter<R> {
+
+  /**
+   * Build table service related commit metadata, 
+   * Especially `EarliestCommitToReta` column when enable incremental table service.
+   * @return metadata to commit
+   */
+  R buildCommitMeta();
+}
+```
+
+Use `TableServiceBaseActionExecutor` to control the behavior of getting partitions to be performed into current table service.
+
+```java
+package org.apache.hudi.table.action;
+
+import org.apache.hudi.common.engine.HoodieEngineContext;
+import org.apache.hudi.common.model.TableServiceType;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieNotSupportedException;
+import org.apache.hudi.table.HoodieTable;
+
+import java.io.IOException;
+import java.util.List;
+
+public abstract class TableServiceBaseActionExecutor<T, I, K, O, R> extends BaseActionExecutor<T, I, K, O, R> implements TableServiceCommitter<R> {
+
+  public TableServiceBaseActionExecutor(HoodieEngineContext context, HoodieWriteConfig config, HoodieTable<T, I, K, O> table, String instantTime) {
+    super(context, config, table, instantTime);
+  }
+
+  /**
+   * Returns the earliest commit to retain and will record in meta as new EarliestCommitToRetain
+   */
+  public Option<HoodieInstant> getEarliestCommitToRetain() {
+    return null;
+  }
+
+  /**
+   * get partitions to be performed in current table service
+   * @param instantToRetain
+   * @param type
+   * @return
+   * @throws IOException
+   */
+  public List<String> getPartitionPaths(Option<HoodieInstant> instantToRetain, TableServiceType type) {
+    return null;
+  }
+
+  /**
+   * Get incremental partitions from EarliestCommitToRetain to instantToRetain
+   * @param instantToRetain
+   * @param type
+   * @param deleteEmptyCommit
+   * @return
+   * @throws IOException
+   */
+  private List<String> getIncrementalPartitionPaths(Option<HoodieInstant> instantToRetain, TableServiceType type) {
+    return null;
+  }
+
+  /**
+   * Get all table partitions related to current table
+   * @param instantToRetain
+   * @param type
+   * @param deleteEmptyCommit
+   * @return
+   * @throws IOException
+   */
+  private List<String> getAllPartitionPaths(Option<HoodieInstant> instantToRetain, TableServiceType type) {
+    return null;
+  }
+
+  public R buildCommitMeta() {
+    return null;
+  }
+
+}
+
+```
+
+### Optimize the inheritance relationship of Table Service
+
+![NewTableServiceDiagram.png](NewTableServiceDiagram.png)
+
+### Work Flow for Incremental Table Service
+
+Table Service Planner
+1. Retrieve the instant recorded in the last completed table service commit as **INSTANT 1**.
+2. Calculate the current instant to be processed as **INSTANT 2**.
+3. Obtain all partitions involved from **INSTANT 1** to **INSTANT 2** as incremental partitions and perform the table service plan operation.
```

Comment:
If we turn on the incremental table service mode, are the various flexible partition selection mechanisms now unavailable? Consider the following scenario: 
- on ts_0, write to two partitions: p_1 and p_2
- on ts_1, a schedule compaction compaction with parittion-selection-strategy that only compacts p_2
- on ts_2, write to p_2 again.
- on ts_3  compaction will only process partitions written between ts_1 and ts_3. Compaction will still only merge p_2. When can a compaction occur that compacts p_1?

Reply:
Also in `IncrementalxxxxStrategy` maybe we can record `missing partitions` in plan and Process the missing partitions together with the new fetched incremental partitions next time